barman/server.py

```diff
@@ -1515,6 +1515,60 @@ def show(self):
     def delete_backup(self, backup):
         """Deletes a backup
 
+        :param backup: the backup to delete
+        """
+        if self.backup_manager.should_keep_backup(backup.backup_id):
+            output.warning(
+                "Skipping delete of backup %s for server %s "
+                "as it has a current keep request. If you really "
+                "want to delete this backup please remove the keep "
+                "and try again.",
+                backup.backup_id,
+                self.config.name,
+            )
+            return
+
+        # Honour minimum required redundancy
+        available_backups = self.get_available_backups(status_filter=(BackupInfo.DONE,))
+        minimum_redundancy = self.config.minimum_redundancy
+        if backup.status == BackupInfo.DONE and minimum_redundancy >= len(
+            available_backups
+        ):
+            output.warning(
+                "Skipping delete of backup %s for server %s "
+                "due to minimum redundancy requirements "
+                "(minimum redundancy = %s, "
+                "current redundancy = %s)",
+                backup.backup_id,
+                self.config.name,
+                minimum_redundancy,
+                len(available_backups),
+            )
+            return
+
+        if backup.children_backup_ids:
+            output.info(
+                "Backup %s has incremental backups which depent on it. "
+                "Deleting all backups in the tree",
+                backup.backup_id,
+            )
+
+        # Deletes the backup along with all its descendants in the
+        # backup tree i.e. all its subsequent incremental backups.
+        # If it has no descendants, only the current backup is deleted.
+        # NOTE: The concept of backup tree only applies when dealing with Postgres
+        # incremental backups, since every rsync backup is complete on its own and
+        # does not hold any dependency relation with other backups in the catalog
+        deleted = False
+        backups_to_delete = backup.walk_backups_tree()
+        for backup in backups_to_delete:
+            deleted = self.perform_delete_backup(backup)
```

Comment:
I need to take a closer look at this PR.
IAC I wonder that this `deleted` variable will hold only the value of the "root" backup deletion.
What should we do if any child fails to be deleted, e.g. if we can't acquire a lock? At a first glance it seems we would have a leftover.

Reply:
FWIW by coincidence I hit an issue more or less like the one I brought up above. After that I was never able to run `barman delete` and had to physically remove the files from `barman_home/barman_server/base` directory to recover from that:

```console
(barman) [vagrant@barmandevhost barman]$ barman list-backup pg17
pg17 20240624T215009 'child2.1' - Mon Jun 24 19:50:10 2024 - Size: 35.7 MiB - WAL Size: 0 B
pg17 20240624T214947 'child2' - Mon Jun 24 19:49:48 2024 - Size: 35.7 MiB - WAL Size: 64.0 MiB
pg17 20240624T214834 'child1.1.2' - Mon Jun 24 19:48:35 2024 - Size: 35.7 MiB - WAL Size: 112.0 MiB
pg17 20240624T214828 'child1.1.1' - Mon Jun 24 19:48:29 2024 - Size: 35.7 MiB - WAL Size: 32.0 MiB
pg17 20240624T214819 'child1.1' - Mon Jun 24 19:48:21 2024 - Size: 35.7 MiB - WAL Size: 32.0 MiB
pg17 20240624T214809 'child1' - Mon Jun 24 19:48:10 2024 - Size: 35.7 MiB - WAL Size: 32.0 MiB
pg17 20240624T214757 'root_backup' - Mon Jun 24 19:48:00 2024 - Size: 355.7 MiB - WAL Size: 32.0 MiB
(barman) [vagrant@barmandevhost barman]$ barman delete pg17 root_backup
Backup 20240624T214757 has incremental backups which depent on it. Deleting all backups in the tree
Deleting backup 20240624T214828 for server pg17
Deleted backup 20240624T214828 (start time: Mon Jun 24 21:50:59 2024, elapsed time: less than one second)
Deleting backup 20240624T214834 for server pg17
Deleted backup 20240624T214834 (start time: Mon Jun 24 21:50:59 2024, elapsed time: less than one second)
Deleting backup 20240624T214819 for server pg17
Deleted backup 20240624T214819 (start time: Mon Jun 24 21:50:59 2024, elapsed time: less than one second)
Deleting backup 20240624T214809 for server pg17
Deleted backup 20240624T214809 (start time: Mon Jun 24 21:50:59 2024, elapsed time: less than one second)
Deleting backup 20240624T214843 for server pg17
EXCEPTION: Could not find backup_id 20240624T214843
See log file for more details.
(barman) [vagrant@barmandevhost barman]$ barman list-backup pg17
(barman) [vagrant@barmandevhost barman]$ barman list-backup pg17
pg17 20240624T215009 'child2.1' - Mon Jun 24 19:50:10 2024 - Size: 35.7 MiB - WAL Size: 0 B
pg17 20240624T214947 'child2' - Mon Jun 24 19:49:48 2024 - Size: 35.7 MiB - WAL Size: 64.0 MiB
pg17 20240624T214757 'root_backup' - Mon Jun 24 19:48:00 2024 - Size: 355.7 MiB - WAL Size: 240.0 MiB
(barman) [vagrant@barmandevhost barman]$ barman delete pg17 root_backup
Backup 20240624T214757 has incremental backups which depent on it. Deleting all backups in the tree
Deleting backup 20240624T214809 for server pg17
EXCEPTION: Could not find backup_id 20240624T214809
See log file for more details.
(barman) [vagrant@barmandevhost barman]$ barman delete pg17 20240624T215009
Deleting backup 20240624T215009 for server pg17
Deleted backup 20240624T215009 (start time: Mon Jun 24 21:52:31 2024, elapsed time: less than one second)
(barman) [vagrant@barmandevhost barman]$ barman list-backup all
pg17 20240624T214947 'child2' - Mon Jun 24 19:49:48 2024 - Size: 35.7 MiB - WAL Size: 64.0 MiB
pg17 20240624T214757 'root_backup' - Mon Jun 24 19:48:00 2024 - Size: 355.7 MiB - WAL Size: 240.0 MiB
(barman) [vagrant@barmandevhost barman]$ barman delete pg17 20240624T214947
Backup 20240624T214947 has incremental backups which depent on it. Deleting all backups in the tree
Deleting backup 20240624T214952 for server pg17
EXCEPTION: Could not find backup_id 20240624T214952
See log file for more details.
(barman) [vagrant@barmandevhost barman]$ barman list-backup all
pg17 20240624T214947 'child2' - Mon Jun 24 19:49:48 2024 - Size: 35.7 MiB - WAL Size: 64.0 MiB
pg17 20240624T214757 'root_backup' - Mon Jun 24 19:48:00 2024 - Size: 355.7 MiB - WAL Size: 240.0 MiB
(barman) [vagrant@barmandevhost barman]$ barman delete pg17 20240624T214757
Backup 20240624T214757 has incremental backups which depent on it. Deleting all backups in the tree
Deleting backup 20240624T214809 for server pg17
EXCEPTION: Could not find backup_id 20240624T214809
See log file for more details.
(barman) [vagrant@barmandevhost barman]$ barman list-backup all
pg17 20240624T214947 'child2' - Mon Jun 24 19:49:48 2024 - Size: 35.7 MiB - WAL Size: 64.0 MiB
pg17 20240624T214757 'root_backup' - Mon Jun 24 19:48:00 2024 - Size: 355.7 MiB - WAL Size: 240.0 MiB
```

Reply:
Maybe we can treat this whole deletion of a chain as a single unit, something like transactions in postgres, all or nothing.

Reply:
> I need to take a closer look at this PR. IAC I wonder that this `deleted` variable will hold only the value of the "root" backup deletion. What should we do if any child fails to be deleted, e.g. if we can't acquire a lock? At a first glance it seems we would have a leftover.

That's something I was thinking as well. I mean, the errors will be logged so the user will know which backups couldn't be deleted and could potentially delete them manually. On the other hand, if we stop the loop in case any backup fails along the way it would also require the user to manually fix whatever is wrong and try again. Both requires a manual intervention, but thinking more about it, maybe it's better to force him to delete everything before deleting the parent to avoid leftovers, as you said.

Reply:
@barthisrael I think this error that you faced is because we are not deleting updating the children_ids in the backup_info as children are deleted. So when you try to run the delete for the second time it tries to find backups that were already deleted but still referenced in the parent file. That's something important that I was not even remembering: update the references as backups are deleted 😵

Reply:
> Maybe we can treat this whole deletion of a chain as a single unit, something like transactions in postgres, all or nothing.

that would be great, I just don't have any idea on how to accomplish that :/

Reply:
Pushed some changes to stop the process if an incremental fails to be deleted and also a change to update the children_ids they are deleted.

Reply:
> I need to take a closer look at this PR. IAC I wonder that this `deleted` variable will hold only the value of the "root" backup deletion. What should we do if any child fails to be deleted, e.g. if we can't acquire a lock? At a first glance it seems we would have a leftover.

The lock issue you raised is interesting, and I think that we should look into locking sequence and be sure that a lock is acquired and kept for the whole chain-delete action. avoiding other interactions from external processes.
I'm just thinking out loud here btw.

Reply:
> Maybe we can treat this whole deletion of a chain as a single unit, something like transactions in postgres, all or nothing.

This is potentially a good idea if we can implement something like a dry run of the deletion of the whole chain...
Otherwise can potentially become something that is going to consume a lot of disk space...

Reply:
> The lock issue you raised is interesting, and I think that we should look into locking sequence and be sure that a lock is acquired and kept for the whole chain-delete action. avoiding other interactions from external processes. I'm just thinking out loud here btw.

Do you think we need a lock on every single backup beforehand or should a lock on the server be enough? I was thinking of acquiring a server level lock on the start and then procede to acquire backup level locks on demand as incrementals are being deleted. What do you guys think? @gcalacoci @barthisrael @andremagui

Reply:
I would like to know more about locking to make meaningful suggestions :D but i feel that acquiring just one lock would be enough. As it is, in the start of delete_backup method in the server there is a check if its possible to acquire a server lock and then releases it, but in fact just one lock is created when deleting backups just before the call to backup_manager. So does it make sense to have one lock for all deletions or a lock for each deletion? 

Sorry, instead of helping i just added more questions!!!!